### ISSUE TITLE
About the  SoftDiceLoss loss in nnUNetTrainer, Thx~

### ISSUE THREAD

**Issue by AllenPyPy** [MIC-DKFZ/nnUNet#108]
Hi FabianIsensee,
Thank you for your great great great work of this nnUNet! This framework helps me a lot and it always has such a great result in medical image segmenation! :）

I'm now learning your code but met some problem. In your code:

In initialize_network_optimizer_and_scheduler(), the self.network = Generic_UNet(...,final_nonlin = lambda x: x,...), so that the Generic_UNet will have a result with logits(or whatever, kernel or conv =1 without nonlin layer), right?

https://github.com/MIC-DKFZ/nnUNet/blob/16003dc00374e4ccac221be787a788bf47ca6e62/nnunet/training/network_training/nnUNetTrainer.py#L86

And in nnUNetTrainer.py, you set the params of DC_and_CE_loss() as above and it means the apply_nonlin=None in SoftDiceLoss function, right? 

So, the dice loss of your work will be calculated with logits and y_onehot, right?

But I thought we generally get dice loss with softmax or sigmoid result(probability) , isn't it? So when I just use the ori code, I will get a weird dice loss cause the logits may be negative value. What are the benefits of this design (get dice loss with loguts) ?

Many Thanks~~~

**Comment by FabianIsensee** [comment]
https://github.com/MIC-DKFZ/nnUNet/blob/16003dc00374e4ccac221be787a788bf47ca6e62/nnunet/training/loss_functions/dice_loss.py#L119

nonlineartiy is always set to softmax for the dice loss :-)

Best,
Fabian

**Comment by AllenPyPy** [comment]
> https://github.com/MIC-DKFZ/nnUNet/blob/16003dc00374e4ccac221be787a788bf47ca6e62/nnunet/training/loss_functions/dice_loss.py#L119
> 
> (taken from DC_and_CE_loss implementation)
> 
> nonlineartiy is always set to softmax for the dice loss :-)
> 
> Best,
> Fabian

Oh, I see, I missed the apply_nonlin and just used SoftDiceLoss, so stupid am I. Thank you~
